smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpBindingProtocolGenerator.java

```diff
@@ -820,6 +944,124 @@ private void readHeaders(
+                            writer.write("{} as {[key: string]: any}");
+                        });
+                writer.openBlock("const eventMessage = {", "};", () -> {
+                    writer.write("headers: eventHeader,");
+                    writer.write("body: event[eventName].body");
+                });
+                writer.openBlock("const parsedEvent = {", "};", () -> {
+                    writer.write("[eventName]: eventMessage");
+                });
+                Symbol targetSymbol = context.getSymbolProvider().toSymbol(target);
+                if (target instanceof StructureShape) {
+                    // Single-event stream. Save the structure and generate event-specific deser later.
+                    this.deserializingEventShapes.add(target.asStructureShape().get());
+                    // For single-event stream, supply event message to corresponding event structure deser.
+                    String deserFunction = ProtocolGenerator.getDeserFunctionName(
+                            targetSymbol, context.getProtocolName()) + "_event(eventMessage, context)";
+                    writer.write("return await $L;", deserFunction);
+                } else if (target instanceof UnionShape) {
+                    // Multi-event stream. Save the union and generate dispatcher later.
+                    this.deserializeEventUnions.add(target.asUnionShape().get());
+                    // For multi-event stream, supply event name to event pairs to the events union deser.
+                    String deserFunction = ProtocolGenerator.getDeserFunctionName(
+                            targetSymbol, context.getProtocolName()) + "_event(parsedEvent, context)";
+                    writer.write("return await $L;", deserFunction);
+                }
+            });
+        });
+    }
+
+    private void generateEventDeserializer(
+            GenerationContext context,
+            StructureShape event
+    ) {
+        TypeScriptWriter writer = context.getWriter();
+        SymbolProvider symbolProvider = context.getSymbolProvider();
+        Symbol symbol = symbolProvider.toSymbol(event);
+        String methodName = ProtocolGenerator.getDeserFunctionName(symbol, context.getProtocolName()) + "_event";
+        Model model = context.getModel();
+        // Handle the general response.
+        writer.openBlock("const $L = async (\n"
+                        + "  output: any,\n"
+                        + "  context: __SerdeContext\n"
+                        + "): Promise<$T> => {", "}",
+                methodName,
+                symbol, () -> {
+                    writer.openBlock("let contents: $L = {", "} as any;", symbol.getName(), () -> {
+                        if (!event.getAllMembers().values().isEmpty()) {
+                            writer.write("__type: $S,", event.getId().getName());
+                        }
+                    });
+                    // Parse members from event headers.
```

Comment:
Can these chunks of code for handling message headers and bodies be broken down in to other functions like for standard http responses.